File: src-backend/models/account.py
from typing import Optional

from sqlalchemy import Boolean, Enum, ForeignKey, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from core.database import Base
from enums.account_type_enum import AccountType
from enums.currency_enum import Currency


class Account(Base):
    __tablename__ = "accounts"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(Text, nullable=False)
    account_type: Mapped[AccountType] = mapped_column(Enum(AccountType), nullable=False)
    currency: Mapped[Currency] = mapped_column(Enum(Currency), nullable=False)
    is_active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    linked_cash_account_id: Mapped[Optional[int]] = mapped_column(
        ForeignKey("accounts.id"), nullable=True
    )

    linked_cash_account: Mapped[Optional["Account"]] = relationship(
        "Account", remote_side="Account.id", foreign_keys=[linked_cash_account_id]
    )
    transactions: Mapped[list["Transaction"]] = relationship( # type: ignore
        "Transaction", back_populates="account"
    )
    assets: Mapped[list["Asset"]] = relationship( # type: ignore
        "Asset", back_populates="account"
    )

File: packages/shared/src/schemas/account.ts
import { z } from 'zod'
import { CurrencyEnum } from '../enums/currency.js'

export const AccountTypeEnum = z.enum(['cash', 'investment'])
export type AccountType = z.infer<typeof AccountTypeEnum>

export const AccountSchema = z.object({
  id:                     z.number(),
  name:                   z.string(),
  account_type:           AccountTypeEnum,
  currency:               CurrencyEnum,
  linked_cash_account_id: z.number().nullable(),
  is_active:              z.boolean(),
})
export type Account = z.infer<typeof AccountSchema>

// is_active is set automatically (true at first, false on delete),
// so the client never sends it.
export const NewAccountSchema = AccountSchema
  .omit({ id: true, is_active: true })
  .extend({ linked_cash_account_id: z.number().nullable().optional() })
export type NewAccount = z.infer<typeof NewAccountSchema>
